extras/indexing/src/main/java/org/apache/rya/indexing/smarturi/duplication/DuplicateDataDetector.java

```diff
@@ -0,0 +1,1066 @@
+        return allValuesNearlyEqual;
+    }
+
+    /**
+     * Gets the appropriate {@link ApproxEqualsDetector} for the specified
+     * class.
+     * @param clazz the {@link Class} to find an {@link ApproxEqualsDetector}
+     * for.
+     * @return the {@link ApproxEqualsDetector} for the class or {@code null} if
+     * none could be found.
+     */
+    public ApproxEqualsDetector<?> getDetectorForType(final Class<?> clazz) {
+        return classMap.get(clazz);
+    }
+
+    private static boolean isOnlyOneNull(final Object lhs, final Object rhs) {
+        return (lhs == null && rhs != null) || (lhs != null && rhs == null);
+    }
+
+    /**
+     * Class to detect if two booleans are considered approximately equal to
+     * each other.
+     */
+    public static class BooleanApproxEqualsDetector implements ApproxEqualsDetector<Boolean> {
+        private static final Tolerance DEFAULT_TOLERANCE = new Tolerance(0.0, ToleranceType.DIFFERENCE);
+        private final Tolerance tolerance;
+
+        /**
+         * Creates a new instance of {@link BooleanApproxEqualsDetector}.
+         * @param tolerance the {@link Tolerance}.
+         */
+        public BooleanApproxEqualsDetector(final Tolerance tolerance) {
+            this.tolerance = tolerance != null ? tolerance : getDefaultTolerance();
+        }
+
+        @Override
+        public boolean areObjectsApproxEquals(final Boolean lhs, final Boolean rhs) {
+            // Should never be almost equals when tolerance is 0, only exactly equals
+            // Otherwise if there's any tolerance specified everything is equal
+            return tolerance.getValue() == 0 ? Objects.equals(lhs, rhs) : true;
+        }
+
+        @Override
+        public Tolerance getDefaultTolerance() {
+            return DEFAULT_TOLERANCE;
+        }
+
+        @Override
+        public Boolean convertStringToObject(final String string) throws SmartUriException {
+            return Boolean.valueOf(string);
+        }
+
+        @Override
+        public Class<?> getTypeClass() {
+            return Boolean.class;
+        }
+
+        @Override
+        public URI getXmlSchemaUri() {
+            return XMLSchema.BOOLEAN;
+        }
+    }
+
+    /**
+     * Class to detect if two bytes are considered approximately equal to each
+     * other.
+     */
+    public static class ByteApproxEqualsDetector implements ApproxEqualsDetector<Byte> {
+        private static final Tolerance DEFAULT_TOLERANCE = new Tolerance(0.0, ToleranceType.DIFFERENCE);
+        private final Tolerance tolerance;
+
+        /**
+         * Creates a new instance of {@link ByteApproxEqualsDetector}.
+         * @param tolerance the {@link Tolerance}.
+         */
+        public ByteApproxEqualsDetector(final Tolerance tolerance) {
+            this.tolerance = tolerance != null ? tolerance : getDefaultTolerance();
+        }
+
+        @Override
+        public boolean areObjectsApproxEquals(final Byte lhs, final Byte rhs) {
```

Comment:
Seems like this is Serialization dependent.  Maybe it would be a good idea to include similar method that takes in an implementation of a standard Serialization interface along with two java objects? Or maybe just two Java serializables?  I guess this doesn't work in the context of this particular implementation.  Maybe it would would be good to have an ApproxEqualsDetector class with a Serializable parameter?  That way, you know how things are being serialized.  That could just delegate to this class under the hood.